Connection Filter System Fix

evertn opened the pull request:
### ConnectionController
[(ConnectionController.cs)](https://github.com/voxeltycoon-community/voxel-tycoon-open-library/blob/2e8b76d64bbd172d27ac46687337e1c39efa2b24/Voxel%20Tycoon%20Open%20Library/StorageNetwork/ConnectionManager/ConnectionController.cs)

- Assigned new namespace `VTOL.StorageNetwork.ConnectionManager`.

Changed `RegisterConnectionFilter()` so it is no longer required to register a `IConnectionFilter` with `RegisterConnectionFilter(new IConnectionFilter, priority)`. Instead you can just give the type of the class you want to register:

`RegisterConnectionFilter<TFilter>(double priority = 0);`

This new method will use the old method creating a new instance of `TFilter`.

- Changed the original `RegisterConnectionFilter`-method to private
- Removed summary from original `RegisterConnectionFilter`-method
- Added a new public `RegisterConnectionFilter`-method
- Added summary to the new `RegisterConnectionFilter`-method

### IConnectionFilter

[(IConnectionFilter.cs)](https://github.com/voxeltycoon-community/voxel-tycoon-open-library/blob/2e8b76d64bbd172d27ac46687337e1c39efa2b24/Voxel%20Tycoon%20Open%20Library/StorageNetwork/ConnectionManager/IConnectionFilter.cs)

- Assigned new namespace `VTOL.StorageNetwork.ConnectionManager`.

### InvalidateSiblingsPatch

[(InvalidateSiblingsPatch.cs)](https://github.com/voxeltycoon-community/voxel-tycoon-open-library/blob/2e8b76d64bbd172d27ac46687337e1c39efa2b24/Voxel%20Tycoon%20Open%20Library/StorageNetwork/ConnectionManager/InvalidateSiblingsPatch.cs)

- Assigned new namespace `VTOL.StorageNetwork.ConnectionManager`.
- Changed some wording in the summary
- Turned out there was a `boolean` tracking if a building was built: `Building.IsBuilt`. Using this boolean to check if a building is built instead of using `Building.Id`.
- Updated summary so it complies with the new `ConnectionController.RegisterConnectionFilter`-method.

### MarkBuildingDirtyPatch (new)

[(MarkBuildingDirtyPatch.cs)](https://github.com/voxeltycoon-community/voxel-tycoon-open-library/blob/2e8b76d64bbd172d27ac46687337e1c39efa2b24/Voxel%20Tycoon%20Open%20Library/StorageNetwork/ConnectionManager/MarkBuildingDirtyPatch.cs)

It turned out that when a StorageNetworkBuilding was built, the building itself would not update it's Storage Network, only the buildings around it within range did. For the Connection Filter system to work correctly we required the newly built building to update it's Storage Network too.

- Added a prefix to `StorageNetworkBuilding.OnBuilt()` to use `MarkSiblingsDirty()` for the current instance. This will make sure the Storage Network is also updated for the building that has been built in the current frame.

### PotentialConnection

[(PotentialConnection.cs)](https://github.com/voxeltycoon-community/voxel-tycoon-open-library/blob/2e8b76d64bbd172d27ac46687337e1c39efa2b24/Voxel%20Tycoon%20Open%20Library/StorageNetwork/ConnectionManager/PotentialConnection.cs)

- Assigned new namespace `VTOL.StorageNetwork.ConnectionManager`.

### PotentialConnectionArgs

[(PotentialConnectionArgs.cs)](https://github.com/voxeltycoon-community/voxel-tycoon-open-library/blob/2e8b76d64bbd172d27ac46687337e1c39efa2b24/Voxel%20Tycoon%20Open%20Library/StorageNetwork/ConnectionManager/PotentialConnectionArgs.cs)

- Assigned new namespace `VTOL.StorageNetwork.ConnectionManager`.

### PriorityConnectionFilter

[(PriorityConnectionFilter.cs)](https://github.com/voxeltycoon-community/voxel-tycoon-open-library/blob/2e8b76d64bbd172d27ac46687337e1c39efa2b24/Voxel%20Tycoon%20Open%20Library/StorageNetwork/ConnectionManager/PriorityConnectionFilter.cs)

### Folders

- Created a folder for the new namespace

evertn committed:
Connection Filter System Fix

It turned out that when a StorageNetworkBuilding was built, the building itself would not update it's Storage Network, only the buildings around it withing did. For the Connection Filter system to work correctly we required the built building to update it's Storage Network too.

- Assigned new .StorageNetwork.ConnectionManager - namespace too all the classes and interfaces part of the connection filter system.
- Created a folder for this new namespace.

evertn committed:
RegisterConnectionFilter-method update

Changed the RegisterConnectionFilter-method so it is no longer required to register a IConnectionFilter with RegisterConnectionFilter(new IConnectionFilter, priority). Instead you can just give the type of the class you want to register.

- Changed the original RegisterConnectionFilter() to private
- Removed summary from original RegisterConnectionFilter-method
- Added public RegisterConnectionFilter<TFilter>(double priority = 0)
- Added summary to new RegisterConnectionFilter-method

kevin4998 commented on Voxel Tycoon Open Library/StorageNetwork/ConnectionManager/ConnectionController.cs:
```diff
@@ -14,21 +14,34 @@ public class ConnectionController : LazyManager<ConnectionController>
-		/// Registers a method which decides if a connection between two <see cref="StorageNetworkBuilding"/> should be allowed or not.
+		/// Registers a <see cref="IConnectionFilter"/> to the connection filter system.
 		/// </summary>
-		/// <param name="connectionFilter">The class which decides if a connection should be canceled or not.</param>
+		/// <typeparam name="TFilter">Type of class which holds the <see cref="IConnectionFilter"/> fucntionality.</typeparam>
 		/// <param name="priority">(Optional) The priority of the filter. Default value is 0.</param>
```
[Spelling]

fucntionality -> functionality

It might be useful to add a spell check extension to visual studio to assist with writing comments. There seems to be a popular one online called **Visual Studio Spell Checker**:
- [VS 2019 version](https://marketplace.visualstudio.com/items?itemName=EWoodruff.VisualStudioSpellCheckerVS2017andLater)
- [VS 2022 version](https://marketplace.visualstudio.com/items?itemName=EWoodruff.VisualStudioSpellCheckerVS2022andLater)

evertn replied:
Installed

kevin4998 commented on Voxel Tycoon Open Library/StorageNetwork/ConnectionManager/ConnectionController.cs:
```diff
@@ -14,21 +14,34 @@ public class ConnectionController : LazyManager<ConnectionController>
+			{
+				RegisterConnectionFilter(new TFilter(), priority);
+			}
+			catch (InvalidOperationException)
 			{
 				throw new InvalidOperationException($"You are not allowed to register after state OnGameStarting. The current state is {Vtol.GameState}.");
 			}
```
[Question]

Why do you catch the exception to just throw it again with a message? Why not let the lower exception pass through this method? That way there is only one exception with the message we need. We should only catch exceptions if we can handle them or if we can provide more information about it.

evertn replied:
Removed the `Catch`.

kevin4998 commented on Voxel Tycoon Open Library/StorageNetwork/ConnectionManager/MarkBuildingDirtyPatch.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace VTOL.StorageNetwork.ConnectionManager
+{
+	/// <summary>
+	/// This patch makes sure that a <see cref="StorageNetworkBuilding"/> updates it's Storage Network in the same frame it has been built.
+	/// <para>After some investigating we found that a <see cref="StorageNetworkBuilding"/> is not updating it's Storage Network in the frame it has been build. Instead it would only update whenever another <see cref="StorageNetworkBuilding"/> is build within range.
+	/// For the Connection Filters to work correctly it is necessary that the network is updated after building too. This patch makes sure this happens.</para>
+	/// </summary>
```
[Compliment]

Nice and clear description of why we made this, good work! The only improvement I would come up with is to try and explain why the Connection Filters need this change. Currently you describe it as "to work correctly". But it might be useful to mention that it will seek those connections which get formed when the building is checking it surroundings.

evertn replied:
Tried to write a clearer explanation.

kevin4998 reviewed (dismissed):
Everything is clear and no errors get given by Visual Studio. I only had some minor points but this fix is fine on its own!

evertn committed:
Review requested changes

- Removed the error-handling in the RegisterConnectionFilter()-method
- Updated the summary of MarkBuildingDirtyPatch
- Corrected "fucntionality" to "functionality"

evertn committed:
Summary corrections
